### ISSUE TITLE
[LibOS] Elf relocations do not work correctly for small base addresses

### ISSUE THREAD

**Issue by boryspoplawski** [oscarlab/graphene#1408]
## Description of the problem:
If a binary will be relocated to a small address and some of its segments have offsets bigger than base address, they won't be relocated:
https://github.com/oscarlab/graphene/blob/fe123b7f94115d821fd046b6127a0c4cefc8f85e/LibOS/shim/src/elf/rel.h#L20-L22
For example, lets take `Runtime/ld-linux-x86-64.so.2` and:
- due to aslr base address is randomized to `0x1d8000`
- this elf has 2 LOAD segments - the second is to be loaded at offset `0x224000` from the first
Now since `0x224000 > 0x1d8000` this segment will be loaded at address `0x224000` instead of `0x1d8000 + 0x224000 = 0x3fc000` due to bugged code linked above.

## Steps to reproduce:
Run a small binary (eg. just `int main(void) { return 0; }`) with SGX and this patch (probably need to be applied manually - offsets in file might differ):
```
diff --git a/LibOS/shim/src/bookkeep/shim_vma.c b/LibOS/shim/src/bookkeep/shim_vma.c
index 48aab28..9a35d55 100644
--- a/LibOS/shim/src/bookkeep/shim_vma.c
+++ b/LibOS/shim/src/bookkeep/shim_vma.c
@@ -380,6 +380,7 @@ int init_vma(void) {
         goto out;
     }
     current_heap_top -= ALLOC_ALIGN_DOWN(rand % addr_rand_size);
+    current_heap_top = (void*)0x2bb4000;
 #endif

```

**Comment by dimakuv** [comment]
Unknowingly found the same issue while debugging #1441. Indeed there is a problem with a "avoid-double-relocation" hack in `elf/rel.h`.

To fix #1441, I simply changed the predefined address for PIE binaries to be 0x555555554000 (as Linux does), see https://github.com/oscarlab/graphene/pull/1443.

**Comment by pwmarcz** [comment]
#2161 fixes the `RELOCATED` macro mentioned here, and, I think, fixes this bug.

The code has changed, but I did the following on master:

```diff
--- a/LibOS/shim/src/bookkeep/shim_vma.c
+++ b/LibOS/shim/src/bookkeep/shim_vma.c
@@ -606,6 +606,8 @@ int init_vma(void) {
         }
     }
 
+    g_aslr_addr_top = (void*)0x500000;
+
     /* We need 1 vma to create the memmgr. */
     if (!add_to_thread_vma_cache(&init_vmas[0])) {
         debug("Failed to add tmp vma to cache!\n");
```

When running the `bootstrap` executable from regression tests (without SGX), This maps `ld.so` at 0x93000 and produces a segfault in `ld.so`. On #2161, the same example works.
